# Tech Talent Recruiting with Regex

![Cover Image](image.jpg)
("image.jpg")

#### This project was completed in October, 2025.

## 📖 The Background
As a data analyst at a top global HR consultancy, you have the important task: to search through a vast resumes dataset and find candidates for tech jobs.  
This project is not just a chance for you to apply your skills in regular expressions and data preprocessing, but it also plays a pivotal role in matching suitable candidates with the right jobs.

As a Data Analyst at a leading global HR consultancy, your mission is to delve into an extensive database of resumes to identify suitable candidates for tech-focused roles. This task involves using regular expressions to extract key data points and applying data preprocessing techniques to organize this information effectively.

## 💾 The Data

| Column      | Data Type | Description                                                  |
|-------------|-----------|--------------------------------------------------------------|
| `ID`        | float     | Unique identifier for each resume.                           |
| `Resume_str`| object    | Full text of the resume, rich with details for analysis.     |
| `Category`  | object    | Job category of the resume, indicating the field of expertise. |

# Objectives
Apply regular expressions and data preprocessing skills to efficiently organize the `resumes` data.
1) Generate a new DataFrame named `candidates_df` containing the following columns:
- `id`: Candidate ID.
- `job_title`: Most recent job title.
- `tech_skills`: List of technical skills that may include Python, SQL, R, and Excel.
- `education`: Highest educational degree such as PhD, Master, or Bachelor.

⚠️ Keep only records that are complete and do not contain null or empty string values across these columns!

# Summary of Findings
This project utilizes preprocessing techniques to analyze text data & extract particular pieces of information that can help locate suitable candidates for tech-focused roles. The database available provided a vast number of resumes to parse. As such, it was necessary to utilize regular expressions & other preprocessing techniques to retrieve the desired data & transform it into a more organized dataset for the purposes of analysis & more thorough hiring processes.

After the resume data was imported, it was explored at a basic level to determine what kind of problems & inconsistencies were present. Knowing such information was informative in understanding what preprocessing steps were necessary in addition to how the variables might need to be manipulated in order to retrieve the desired information.  
Fortunately, this brief evaluation of the 1,352 candidate records did not find any significant impediments that required initial preprocessing.

### Analysis I - Investigating, Compiling, Organizing Resume Data
In order to obtain the desired information, the provided dataset had to be evaluated to learn how to do so. Retrieving the candidate ID was simple given that it was present in the original dataset. When it came to finding candidates' most recent job title, relevant technical skills, & the highest degree they obtained, the resume data had to be parsed fairly liberally.

##### Investigating the Resumes
The resume variable, named 'Resume_str,' was investigated extensively to determine how to retrieve the desired information; however, given that there were 1,352 resumes, analyzing them all was unreasonable. As such, a few dozen random resumes were sampled & evaluated for commonalities & differences that might make obtaining the desired information easier. Ultimately, this examination revealed that there is an overall general format to this variable; however, parsing it & easily extracting the desired data would not be simple.  
Generally, a resume begins with a fully capitalized job title, some of which are vague (e.g. "SALES"), followed by an indent. Furthermore, some resumes contained unusual characters within these fully capitalized words/word sequences such as "\n" (indicating a new line). This information was taken to refer to a candidate's most recent job title.

Following this first detail, the resumes then generally listed sections of text. They begin with a section header that is delineated with indents before & after. After a section header, there is generally text information pertaining to it. When a section ended, there is typically another indent, a new section header, another indent, & then the associated information for the new section. This pattern generally consists through the rest of a resume.  
Despite this, there are multiple inconsistencies within this general format. Firstly, the section headers were often organized differently throughout the resumes. Additionally, some sections were named differently despite referring to the same information. For example, one resume might name a section "Highlights" whereas another would call it "Skill Highlights." On top of this, some resumes used the same section header multiple times even though each contained different information.

There are several other inconsistencies that were found in the handful of resumes examined that prevent any single preprocessing technique from being applied. For example, even though sections were typically split with indents, indents were also present between text data within sections.

Ultimately, the objectives in this project defined describe specific variables including pieces of information to look for within the resumes. Given that the two remaining variables look for particular details in a resume, regular expressions ("regex" patterns) were utilized to look for & extract such patterns.

##### Compiling the Desired Information
To apply regular expressions, functions from the `re` module were used. As mentioned previously, the candidate ID was obtained straight from the original dataset. The most recent job title was also established to be the fully capitalized text at the beginning of each resume. As such, the `re.search()` function was used to search for such text; the specific pattern can be found in **Analysis I**.

For the technical skills, `re.findall()` was used to look for mentions of Python, Excel, R &/or SQL in a resume regardless of capitalization. If other technical skills become relevant or requisites for certain tech jobs, they could be added to this pattern.

Next, resumes were parsed to look for any educational degrees obtained by a candidate using the `re.findall()` function. To form these lookups, the following degrees & education levels were used: PhD, Doctoral, Masters, Bachelors, Associates, High School. Others could be added to this pattern if relevant.  
Many resumes listed multiple educational accomplishments. To determine the highest level obtained, additional processing had to be employed. Firstly, the order of these levels had to be established. In this project, this educational hierarchy was defined, from highest to lowest, as: Doctoral/PhD, Master/Masters/MCS, Bachelor/Bachelors/BCS, Associates, High School.

Another requirement in the analysis involved eliminating resumes that did not have information pertaining to a candidate's most recent job title, technical skills, or educational accomplishments. The application of this rule truncated the number of suitable candidates from 1,352 to 444. This missingness could be attributed to a candidate not listing the relevant information in their resume either because they did not possess/achieve it or because they simply did not think to include it.

By applying regular expressions & other coding constructs, like for loops, the desired information was obtained & neatly organized into a table; `candidates_df`. If additional requisites for tech-focused roles want to be introduced in hiring processes, additional patterns can be built to find fitting candidates.

For the purposes of transparency, it is possible that the examinations of the resumes performed in this project did not completely identify all potential patterns in the resume data that might dictate how such information was extracted. For example, there might be some additional educational accomplishments & degrees that are not currently detected based on the regex patterns used.  
As noted earlier & in **Analysis I**, not all of the 1,352 resumes were examined. Instead, an evaluation of a couple dozen resumes required assumptions to be made, particularly that all the resumes followed the same general format described previously.

## Importing, Exploratory Data Analysis, Preprocessing
The dataset needs to be imported & investigated to determine what preprocessing steps are necessary. Furthermore, a basic assessment of the dataset can gain an elementary understanding of it & potentially reveal particular intricacies or patterns.

In [22]:
import numpy as np
import pandas as pd
import re

In [23]:
#Import the data
orig_df = pd.read_csv('resumes.csv')
#Inspect the data
display(orig_df.head(5))
display(orig_df.info())

,ID,Resume_str,Category
0,16852973.0,HR ADMINISTRATOR/MARKETING ASSOCIATE\...,HR
1,22323967.0,"HR SPECIALIST, US HR OPERATIONS ...",HR
2,33176873.0,HR DIRECTOR Summary Over 2...,HR
3,27018550.0,HR SPECIALIST Summary Dedica...,HR
4,17812897.0,HR MANAGER Skill Highlights ...,HR


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1352 entries, 0 to 1351
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          1352 non-null   float64
 1   Resume_str  1352 non-null   object 
 2   Category    1352 non-null   object 
dtypes: float64(1), object(2)
memory usage: 31.8+ KB


None

In [24]:
#Inspect values across variables
display(orig_df.iloc[:,2].value_counts(dropna=False).sort_index())

Category
ADVOCATE                  115
AGRICULTURE                63
AUTOMOBILE                 35
BPO                        21
BUSINESS-DEVELOPMENT      119
CHEF                        9
CONSULTANT                115
DESIGNER                  106
DIGITAL-MEDIA              95
FITNESS                   117
HEALTHCARE                113
HR                        108
INFORMATION-TECHNOLOGY    119
SALES                     115
TEACHER                   102
Name: count, dtype: int64

The dataset has 1,352 entries & no missing values. Each variable was explored briefly to get a better understanding of their values.
- **ID**:
    - No missing values. ID's are between 10,005,171.0 - 99,714,410.0.
- **Resume_str:**
    - No missing values. This variable is where most of the analysis (preprocessing) will be. It can contain a whole mess of information & can be very long. It essentially wraps a resume into a single data point by combining multiple pieces of text, dates, etc. into a single data point. There will naturally be inconsistencies in this variable, but these will be explored in the main analyses.
- **Category:**
    - No missing values. There are 15 unique categories each of which accounts for between nine to 119 resumes. There don't appear to be any typos, entry errors, or inconsistencies across these values, so no cleaning will be necessary for this variable.
    - The categories include: ADVOCATE, AGRICULTURE, AUTOMOBILE, BPO, BUSINESS-DEVELOPMENT, CHEF, CONSULTANT, DESIGNER, DIGITAL-MEDIA, FITNESS, HEALTHCARE, HR, INFORMATION-TECHNOLOGY, SALES, TEACHER.

Obviously preprocessing is a major part of this project, but handling other problems, like missing values, is not currently necessary.

# Analysis I - Investigating, Compiling, Organizing Resume Data
The main analysis of this project has a few subgoals, but it focuses on efficiently organizing the provided resume data. More specifically, information from the given dataset needs to be extracted & transformed into a new one, named `candidates_df`, consisting of the variables described below. Additionally, this dataset should only retain records that are complete & do not contain null or empty string values.
- `id`: Candidate ID.
- `job_title`: Most recent job title.
- `tech_skills`: List of technical skills that may include Python, SQL, R, and Excel.
- `education`: Highest educational degree such as PhD, Master, or Bachelor.


**Plan of action:**
Firstly, the original & new datasets both contain a variable indicating the candidate ID. As such, the new variable can simply be retrieved via the 'ID' variable from the original dataset.

When it comes to the other three variables, including job title, tech skills, & education, these details are contained specifically in the 'Resume_str' variable in the original dataset. To understand how to extract this information, this variable needs to be investigated further. In doing so, it can help determine if there are any commonalities or differences across data points that can help streamline this process. For example, if the resume-string data points are formatted the same, it will make the considerably analyses easier.

In [25]:
#Initiate the dataframe, retrieve the 'ID' variable
candidates_df_orig = pd.DataFrame(data=orig_df['ID'])

#Inspect the 'Resume_str' variable. Explore the resumes of 5 random people.
    #indexes explored: 0, 199, 527, 999, 1240
display(orig_df.loc[999,'Resume_str'])

"         SALES         Summary    I am clearly a loyal and friendly dedicated individual who has an ambition to succeed in any given environment. Although I have extensive experience in the Perishable Meats and seafood industry. I love to learn, and am always up to a challenge whatever the situation. I get along well with others, whilst also working efficiently on my own. I am seeking a position where I can develop and excel while giving my best to an employer.       Skills        literate, MS Excel,Microsoft Word, AS400.              Experience      Company Name     May 2004   to   August 2007     Sales   City  ,   State      Responsible for sales and marketing Fresh meat products and keeping accounts in balance.  Responsible for resets,new store openings.          Company Name     April 1997   to   May 2004     sales and marketing   City  ,   State      Responsible for ,Package meat products and keeping accounts in.  balance.  Responsible for cost effective advertising.  Lines, Sunn

To investigate the resume material, random data points were sampled & looked through. Given that there are 1,352 total data points, assumptions were made about the whole dataset based on an evaluation of a handful.

Upon investigating the 'Resume_str' variable, there is not a clear-cut format in how the relevant variables & details are listed. In five random resumes, this variable contained various subsections that were referred to as: "Summary," "Executive Profile," "Highlights," "Accomplishments," "Experience," "Education," "Skills," "Languages," "Interests," "Additional Information," "Skill Highlights," "Professional Experience," "Core Accomplishments," & more. Across these headings, there are some inconsistencies across various resumes. This makes processing this variable more challenging, so several assumptions will have to be made when applying regular expressions. Nevertheless, there are still some keywords, such as "Education" & "Experience," that will be useful in obtaining the desired information.
- Although it is not listed anywhere, the most recent job title is the fully capitalized position listed at the beginning of every resume.

To illustrate & explain these findings, consider the images ('ex_resume_p1.png', 'ex_resume_p2.png') below which are snapshots of the first resume in the original dataset.

![Example Resume - Part 1](ex_resume_p1.jpg)
("ex_resume_p1.jpg")
![Example Resume - Part 2](ex_resume_p2.jpg)
("ex_resume_p2.jpg")

In this resume above, which has the index 0 in the original dataset, it starts with a fully capitalized job position. Usually, they describe a specific position; however, they are sometimes vague (e.g. "SALES" in index 999). This data point is what is considered to be a candidate's most recent job title, which is one of the variables of interest in this analysis.

After this first detail, this resume lists a "Summary" section which contains traits, past jobs & experiences, etc.; however, not all resumes use this same term (e.g. many others use "Executive Profile" like index 527). This resume then goes on to "Highlights" (other resumes use the term "Skill Highlights").  
This general pattern continues throughout many resumes with various terms for each section. These section headers can be hard to spot when just looking through the data points, but they are often distinguished with extra whitespaces before & after.

When comparing this resume to others, it becomes clear quickly that there is no single format as to how this resume information is listed. Some candidates put more effort into organizing & formatting their resumes, causing vast formatting inconsistencies. Some resumes use different characters when separating the sections in their resumes including fancy bullet points (➢), hyphens, extra whitespaces, new lines (\n), or otherwise. Different resumes use different forms of sentence structure, grammar, punctuation, spelling, capitalization, etc.  
Additionally, the identified sections, many of which are listed prior to the resume snapshots above, do not necessarily follow the same order in the resumes that were examined. On top of this, not all resumes contain the same number or the same sections. For example, the resume above contains the sections Summary, Highlights, Accomplishments, & Experience. Conversely, the resume at index 999 contains the sections Summary, Skills, Experience, & Skills (again).

All of this to say, there are an impossible number of inconsistencies to handle if all resume information were to be digested & properly partitioned into unique variables. Fortunately, this project only focuses on three variables contained in these resumes: the most recent job title (which has already been defined, see above), a list of technical skills, & the highest educational degree that a candidate obtained. Regular expressions (or "regex") can be built to obtain such information, even if no such information is listed in a resume.

Using the `re` module, regular expressions can be defined & built to find the desired information within the many resumes regardless of capitalization, neighboring characters, & other potential formatting inconsistencies.  
The "\b" anchor can be used to match characters at the start or end of a word, which in regex, is typically just a sequence of consecutive characters. This will help to find matches that are surrounded by white spaces, tabs, or punctuation characters (e.g. commas, periods).  
To handle capitalization, all possible matches could be included in the regex pattern. Alternatively, the `flags` parameter in some regex functions can be specified to ignore case sensitivity when matching strings; set flags=re.IGNORECASE.

For the three variables of interest, particulars as to how to extract them from a resume are listed below.

##### Most Recent Job Title
- To find the applicant's most recent job title, a regex pattern needs to be built to encapsulate the first set of words (or word) at the beginning of each resume. In the resumes examined, these words are often fully capitalized. In some cases, this job position contains punctuation & other characters like parentheses. To search for fully capitalized words at the beginning of a string, the "^[A-Z]+" pattern can be used. The plus character is important because it searches for one or more such matches in case multiple words are in the specified position. To match for other cases, various metacharacters could be used like "\s" (whitespaces), "\." (periods), "\," (commas), etc.
    - As an example of what this pattern may encounter, the beginning of the resume at the index 1252 is: "GLOBAL DIGITAL SERVICING ~ DIGITAL ONBOARDING COORDINATOR (CONTRACTOR)".
- Since the `re.search()` function scans a string for the first occurrence of a specified pattern, using this function would be more appropriate to find this information.

##### Technical Skills
- To find listed technical skills, a regex pattern can be built that searches for keywords relating to such skills. Since multiple terms are being searched for, the `re.findall()` function would be more appropriate because it can return all matches. To start, the following skills will be searched for: R, SQL, Python, Excel. Others may be added.
    - Note that this function returns all matches in a list. Even if no matches are found, an empty list is returned.

##### Highest Educational Degree
- To find the highest educational degree achieved by an applicant, a similar process can be used as for the technical skills. By supplying various matches, `re.findall()` can retrieve them all. From there, it can be determined which education is the highest by referencing education levels. To form this basis, the following degrees & education levels will be used: PhD, Doctoral, Masters, Bachelors, Associates, High School. Others may be added.
    - For reference, degrees from highest to lowest follow as: Doctoral/PhD, Masters, Bachelors, Associates, High School, etc.
    - Degrees can also be abbreviated. For example, in technical areas, degrees could include the following: MCS, BCS (Master & Bachelor of Computer Science respectively).
    - Some degrees, or parts of a degree, could show up in a resume for other other purposes. For example, "Associate" is often a term that is part of a job title. Such cases need to be considered & potentially excluded in the regex patterns & search functions.

Moreover, an understanding of the relevant regex functions are important. If no matches are found, `re.search()` will output a "NoneType" object. Conversely, `re.findall()` will output an empty list. When no matches to a regular expression are found, the result needs to be an empty string (""). As such, the code will need to specify this as the result when no matches are found.  
On the other hand, if multiple matches are found, then they should be joined together with a comma & a space; e.g. "skill1, skill2".
- To retrieve the matching substrings from `re.search()`, the `.group()` method can be appended.

On another note, many `re` functions contain a `flags` parameter that can specify whether to search for particular patterns with capitalization in mind or not. For example, setting `flags=re.IGNORECASE` will look for patterns regardless of capitalization. This will be useful when searching for technical skills & degrees.

In [26]:
#Build a regex pattern to look for the most recent job title
pattern_recent_job = r"^\s*([A-Z\.\,\-\~\/\(\)\s]+?)(?=\n|\\n|\s{2,}|$)"
    #the ^ specifies to search at the beginning of the string
    #the \s refers to a whitespace; the * indicates that the preceding character(s) will occur 0 or more times
    #the () are used to indicate the capture group--the substring to be extracted
    #the [] encapsulates character sets, anchors, & characters
    #[A-Z] looks for capital letters
    #\. searches for a period; this is similar for other characters including: ",", "-", "~", "/", "(", ")"
    #+? matches the specified pattern a minimum number of times that it occurs; a "lazy" quantifier
    #?= looks forward to later patterns without capturing them
    #the sequence "\n|\nn" looks for newlines &/or literal "\n" characters
        #the | searches for multiple options; \n or \\n
    #\s{2,} looks for 2 or more whitespaces
    #$ specifies to search at the end of the string

#Build a regex pattern to match technical skills
pattern_tech_skill = r"\b(r|sql|python|excel)\b"
    #the \b searches for matches at the start/end of a word; is a convenient way to search for substrings without directly 
        # specifying surrounding characters that are not wanted

#Build a regex pattern to look for all educational degrees
pattern_educ = r"\b(phd|doctoral|master|masters|MCS|BCS|bachelor|bachelors|associates|high school)\b"
    #need to consider if such strings could come up in a resume in a way that doesn't refer to a person's education.
        #e.g. "associate" could refer to an associate-like position. For this reason, "associate" was excluded.

In [27]:
#Initialize lists to store the extracted information
recent_jobs = []
tech_skills = []
educ_levels = []
#this list will retrieve all education levels & help to find how they are formatted, spelled, etc., giving more insight as to
    # the possible education levels in the resumes & what to look out for. It will also help when it comes to ordering them &
    # finding the highest one achieved within each resume.
uniq_educ_orig = []

#Loop through each resume
for res in orig_df['Resume_str']:
    #RECENT JOB POSITION
    #use the regex pattern to look for a match
    job_match = re.search(pattern_recent_job, res)
    #if a match is found
    if job_match is not None:
        #retrieve the job information
        job_match = job_match.group(1).strip()
    #if no match is found, use an empty string
    else:
        job_match = ""
    #append the result to the 'recent_jobs' list
    recent_jobs.append(job_match)
    
    
    #TECH SKILLS
    #use the regex pattern to look for matches (case insensitive)
    tech_skills_matches = re.findall(pattern_tech_skill, res, flags=re.IGNORECASE)
    #apply title case to each match except for "SQL" which should be fully capitalized
    tech_skills_matches = [i.title() if i != 'SQL' else i.upper() for i in tech_skills_matches]
    #obtain the unique tech skills, put them into a list
    tech_skills_matches = list(set(tech_skills_matches))
    #append the matches to the 'tech_skills' list, each formatted as a comma-separated string
    tech_skills.append(", ".join(tech_skills_matches))
    
    
    #EDUCATION LEVEL(S)
    #use the regex pattern to look for matches
    educ_lev_matches = re.findall(pattern_educ, res, flags=re.IGNORECASE)    
    #apply title case to each match
    educ_lev_matches = [i.title() for i in educ_lev_matches]
    #select the unique education levels, put them into a list
    educ_lev_matches = list(set(educ_lev_matches))
    #obtain the unique education levels
    uniq_educ_orig.append(educ_lev_matches)
    #append the matches to the 'educ_levels' list, each formatted as a comma-separated string
    educ_levels.append(", ".join(educ_lev_matches))

#Check results
#display(recent_jobs)

In [31]:
#Put the resulting lists into a data frame
candidates_df = candidates_df_orig.copy()
candidates_df['job_title'] = recent_jobs
candidates_df['tech_skills'] = tech_skills
candidates_df['educ_levels'] = educ_levels

#Only want records with complete values across all variables. Remove those with missing values
    #This reduces the dataset from 1,352 data points to 444 (~ 32.8%)
candidates_df = candidates_df[
    (candidates_df['job_title'] != '') & (candidates_df['tech_skills'] != '') & (candidates_df['educ_levels'] != '')]

#Check results
display(candidates_df.head(5))

,ID,job_title,tech_skills,educ_levels
2,33176873.0,HR DIRECTOR,Excel,"Bachelor, Master"
4,17812897.0,HR MANAGER,Excel,"Bachelor, Associates"
8,11847784.0,HR SPECIALIST,Excel,Bachelor
9,32896934.0,HR CLERK,R,"Bachelor, Associates"
13,11763983.0,HR GENERALIST,Excel,Bachelors


By filtering for records with complete values across the new variables, it truncated the dataset from 1,352 candidates to 444 which is a loss of about two-thirds of the candidates. In other words, this initial phase in the evaluation of candidates' resumes found that roughly two in every three candidates either did not have the requisite information in their resume or lacked the desired accomplishments or skills. This missingness could be attributed to not having mentioned their most recent job title (or not having one), lacking (or failing to mention) certain tech skills (Excel, R, SQL, Python), or not listing or achieving an educational degree.

In [29]:
#Obtain all the unique matches found in the 'educ_levels' variable
uniq_educ = sorted(list(set([edu for element in uniq_educ_orig for edu in element])))
display(uniq_educ)

['Associates',
 'Bachelor',
 'Bachelors',
 'Bcs',
 'Doctoral',
 'High School',
 'Master',
 'Masters',
 'Mcs',
 'Phd']

Now that the desired information has been extracted & organized into readable & easily retrievable variables, the next step in this process involves determining the highest educational degree achieved in each resume. So far, the 'educ_levels' variable in the `candidates_df` data frame contains all degrees obtained/mentioned within a resume. In many resumes, many terms were captured, so each comma-separated list needs to be evaluated to obtain the highest education level. This could be performed in a multitude of ways including complex sorting methods or for loops.

In the cell directly above, it lists the unique educational levels mentioned among the resumes which include: "Associates," "Bachelor/Bachelors," "Bcs," "Doctoral," "High School," "Master/Masters," "Mcs," & "Phd" (these values have been formatted to title case).  
From a hierarchical standpoint, the order of these degrees, from highest to lowest, is listed below. Given the nature of these strings, there is no easy way to sort them according to their prowess (i.e. alphabetical).
- "Doctoral/PhD," "Master/Masters/MCS," "Bachelor/Bachelors/BCS," "Associates," "High School."

In this case, a nested for loop was utilized. It needs to consider cases where either multiple education-level matches or only a single match was found.

In [30]:
#Want to truncate the 'educ_levels' variable to obtain the highest educational degree achieved. Some degrees can also have
    # cleaner formatting applied (e.g. Mcs -> MCS).
#loop through the dataset by row
for i in candidates_df.index:
    #cases where there are multiple education-level matches must have a comma separating them
    if ", " in candidates_df.loc[i,'educ_levels']:
        #search for Doctoral/PhD matches (highest)
        if 'Doctoral' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'Doctoral'
        elif 'Phd' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'PhD'
        
        #search for Master/Masters/Mcs (2nd highest)
        elif 'Master' in candidates_df.loc[i,'educ_levels'] or 'Masters' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'Masters'
        elif 'Mcs' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'MCS'
        
        #search for Bachelor/Bachelors/BCS (3rd highest)
        elif 'Bachelor' in candidates_df.loc[i,'educ_levels'] or 'Bachelors' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'Bachelors'
        elif 'Bcs' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'BCS'
        
        #search for Associates (4th highest)
        elif 'Associates' in candidates_df.loc[i,'educ_levels']:
            candidates_df.loc[i,'education'] = 'Associates'
        #search for High School (5th highest)
        else:
            candidates_df.loc[i,'education'] = 'High School'
    
    #cases where there is only 1 education-level match -- no changes necessary
    else:
        candidates_df.loc[i,'education'] = candidates_df.loc[i,'educ_levels']

#Check the results (444 rows)
display(candidates_df.sample(10, random_state=10))

,ID,job_title,tech_skills,educ_levels,education
1127,29723311.0,IT CONSULTANT,Excel,"Bachelor, Master, High School",Masters
1078,55097118.0,SALES ASSOCIATE,Excel,"Bachelor, High School",Bachelors
698,98300955.0,"PLATFORM ARCHITECT, HEALTHCARE INCUBATION LAB ...",SQL,Bachelor,Bachelor
624,24412546.0,BUSINESS DEVELOPMENT ASSOCIATE,"Excel, R",Master,Master
596,26278597.0,BUSINESS DEVELOPMENT MANAGER,"Excel, R",Bachelor,Bachelor
916,62994611.0,SOFTWARE DEVELOPER,"Python, SQL","Bachelor, Master",Masters
827,17431847.0,REHABILITATION SPECIALIST / MASSAGE THERAPIST,Excel,"Masters, Bachelor",Masters
450,49475708.0,CUSTOMER ADVOCATE,R,High School,High School
456,14064815.0,FAMILY ADVOCATE,Excel,"Bachelor, Master",Masters
95,27496514.0,HR CUSTOMER SERVICE REPRESENTATIVE,Excel,Master,Master


Now that the highest education level for each resume has been obtained & formatted consistently, the dataset can be submitted. Note that for the final submission, the 'educ_levels' variable, which contains all mentions of possible educational degrees, should be removed from the `candidates_df` dataset beforehand. For clarity, there are 444 candidates in this final dataset.